File: sdk-js/src/manifest.ts
import { SDK_VERSION, satisfiesVersionRange } from './compatibility';
import type {
  ManifestValidationResult,
  PluginConfigField,
  PluginFilesystemPermission,
  PluginManifest,
  PluginPackageDefinitionInput,
  PluginProvider,
  PluginRuntimeLanguage,
} from './types';

const PROVIDERS_BY_LANGUAGE: Record<PluginRuntimeLanguage, PluginProvider[]> = {
  javascript: ['deno'],
  python: ['python'],
};

const ALLOWED_TRIGGERS = new Set([
  'download.queued',
  'download.beforeStart',
  'download.completed',
  'download.failed',
]);

const ALLOWED_CONFIG_INPUT_TYPES = new Set([
  'text',
  'textarea',
  'password',
  'number',
  'boolean',
  'file',
  'directory',
  'select',
  'multi-select',
]);

const ALLOWED_FILESYSTEM_PERMISSIONS = new Set<PluginFilesystemPermission>([
  'fs.plugin.read',
  'fs.plugin.write',
  'fs.payload-file.read',
  'fs.payload-directory.read',
  'fs.payload-directory.write',
  'fs.temp.read',
  'fs.temp.write',
  'fs.user-selected.read',
  'fs.user-selected.write',
]);

function validateConfigFieldDefaultValue(field: PluginConfigField): string | null {
  const value = field.defaultValue;
  if (value === undefined) return null;

  switch (field.inputType) {
    case 'text':
    case 'textarea':
    case 'password':
    case 'file':
    case 'directory':
    case 'select':
      return typeof value === 'string' ? null : `${field.key} defaultValue must be a string.`;
    case 'number':
      return typeof value === 'number' && Number.isFinite(value)
        ? null
        : `${field.key} defaultValue must be a number.`;
    case 'boolean':
      return typeof value === 'boolean' ? null : `${field.key} defaultValue must be a boolean.`;
    case 'multi-select':
      return Array.isArray(value) && value.every((item) => typeof item === 'string')
        ? null
        : `${field.key} defaultValue must be an array of strings.`;
    default:
      return `Unsupported inputType for ${field.key}.`;
  }
}

function validateConfigField(field: PluginConfigField, index: number): string[] {
  const errors: string[] = [];
  const label = field.key?.trim() || `configFields[${index}]`;

  if (!field.key?.trim()) {
    errors.push(`configFields[${index}].key is required.`);
  }
  if (!field.label?.trim()) {
    errors.push(`${label} label is required.`);
  }
  if (!ALLOWED_CONFIG_INPUT_TYPES.has(field.inputType)) {
    errors.push(`${label} has unsupported inputType "${field.inputType}".`);
  }

  const requiresOptions = field.inputType === 'select' || field.inputType === 'multi-select';
  if (requiresOptions) {
    if (!field.options?.length) {
      errors.push(`${label} must declare options for ${field.inputType}.`);
    }
  } else if (field.options?.length) {
    errors.push(`${label} cannot declare options for ${field.inputType}.`);
  }

  if (field.options?.length) {
    const optionValues = new Set<string>();
    for (const option of field.options) {
      if (!option.value?.trim()) {
        errors.push(`${label} has an option with an empty value.`);
      }
      if (!option.label?.trim()) {
        errors.push(`${label} has an option with an empty label.`);
      }
      if (optionValues.has(option.value)) {
        errors.push(`${label} has duplicate option value "${option.value}".`);
      }
      optionValues.add(option.value);
    }
  }

  const defaultValueError = validateConfigFieldDefaultValue(field);
  if (defaultValueError) {
    errors.push(defaultValueError);
  }

  if (
    field.inputType === 'select' &&
    typeof field.defaultValue === 'string' &&
    field.options?.length
  ) {
    if (!field.options.some((option) => option.value === field.defaultValue)) {
      errors.push(`${label} defaultValue must match one declared option.`);
    }
  }

  if (
    field.inputType === 'multi-select' &&
    Array.isArray(field.defaultValue) &&
    field.options?.length
  ) {
    const allowedValues = new Set(field.options.map((option) => option.value));
    for (const item of field.defaultValue) {
      if (!allowedValues.has(item)) {
        errors.push(`${label} defaultValue contains unsupported option "${item}".`);
      }
    }
  }

  const usesNumberBounds =
    field.min !== undefined || field.max !== undefined || field.step !== undefined;
  if (field.inputType !== 'number' && usesNumberBounds) {
    errors.push(`${label} can only use min, max, or step with number fields.`);
  }
  if (field.inputType === 'number') {
    for (const [name, value] of [
      ['min', field.min],
      ['max', field.max],
      ['step', field.step],
    ] as const) {
      if (value !== undefined && (!Number.isFinite(value) || Number.isNaN(value))) {
        errors.push(`${label} ${name} must be a finite number.`);
      }
    }
    if (typeof field.min === 'number' && typeof field.max === 'number' && field.min > field.max) {
      errors.push(`${label} min cannot be greater than max.`);
    }
  }

  return errors;
}

export function slugifyPluginName(input: string): string {
  let slug = '';
  let previousDash = false;

  for (const char of input.trim()) {
    if (/^[a-z0-9]$/i.test(char)) {
      slug += char.toLowerCase();
      previousDash = false;
      continue;
    }

    if (!previousDash) {
      slug += '-';
      previousDash = true;
    }
  }

  const normalized = slug.replace(/^-+|-+$/g, '');
  return normalized || 'plugin';
}

export function getAllowedProviders(language: PluginRuntimeLanguage): PluginProvider[] {
  return [...PROVIDERS_BY_LANGUAGE[language]];
}

export function getManifestValidationErrors(manifest: PluginManifest): string[] {
  const errors: string[] = [];

  if (!manifest.id?.trim()) {
    errors.push('id is required.');
  }

  if (!manifest.slug?.trim()) {
    errors.push('slug is required.');
  }

  if (!manifest.name?.trim()) {
    errors.push('name is required.');
  }

  if (!manifest.version?.trim()) {
    errors.push('version is required.');
  }

  if (typeof manifest.icon === 'string' && manifest.icon.trim() === '') {
    errors.push('icon cannot be empty when provided.');
  }

  if (!manifest.runtime) {
    errors.push('runtime is required.');
    return errors;
  }

  if (!manifest.runtime.entrypoint?.trim()) {
    errors.push('runtime.entrypoint is required.');
  }

  if (!manifest.runtime.supportedProviders?.length) {
    errors.push('runtime.supportedProviders must contain at least one provider.');
  } else {
    const allowedProviders = new Set(getAllowedProviders(manifest.runtime.language));
    for (const provider of manifest.runtime.supportedProviders) {
      if (!allowedProviders.has(provider)) {
        errors.push(
          `runtime.supportedProviders contains unsupported provider "${provider}" for language "${manifest.runtime.language}".`,
        );
      }
    }
  }

  if (
    manifest.runtime.preferredProvider &&
    !manifest.runtime.supportedProviders?.includes(manifest.runtime.preferredProvider)
  ) {
    errors.push('runtime.preferredProvider must be included in runtime.supportedProviders.');
  }

  if (typeof manifest.timeoutSec === 'number' && manifest.timeoutSec <= 0) {
    errors.push('timeoutSec must be greater than 0.');
  }

  if (manifest.permissions && 'env' in (manifest.permissions as Record<string, unknown>)) {
    errors.push(
      'permissions.env is obsolete. Define plugin configuration fields with configFields instead.',
    );
  }

  if (manifest.permissions?.fs?.length) {
    const seenFs = new Set<string>();
    for (const permission of manifest.permissions.fs) {
      if (!ALLOWED_FILESYSTEM_PERMISSIONS.has(permission)) {
        errors.push(`permissions.fs contains unsupported capability "${permission}".`);
      } else if (seenFs.has(permission)) {
        errors.push(`permissions.fs contains duplicate capability "${permission}".`);
      } else {
        seenFs.add(permission);
      }
    }

    const needsUserSelected =
      manifest.permissions.fs.includes('fs.user-selected.read') ||
      manifest.permissions.fs.includes('fs.user-selected.write');
    if (
      needsUserSelected &&
      !manifest.configFields?.some(
        (field) => field.inputType === 'file' || field.inputType === 'directory',
      )
    ) {
      errors.push(
        'permissions.fs uses fs.user-selected.* but configFields does not declare any file or directory inputs.',
      );
    }
  }

  if (!manifest.triggers?.length) {
    errors.push('triggers must contain at least one runtime trigger string.');
  } else {
    for (const trigger of manifest.triggers) {
      if (!ALLOWED_TRIGGERS.has(trigger)) {
        if (trigger.startsWith('triggers.')) {
          errors.push(
            `triggers contains "${trigger}", but plugin.json must use raw runtime names like "download.completed", not SDK identifiers like "triggers.downloadCompleted".`,
          );
        } else {
          errors.push(`triggers contains unsupported runtime trigger "${trigger}".`);
        }
      }
    }
  }

  if (manifest.i18n?.defaultLocale) {
    if (
      manifest.i18n.supportedLocales?.length &&
      !manifest.i18n.supportedLocales.includes(manifest.i18n.defaultLocale)
    ) {
      errors.push('i18n.defaultLocale must be included in i18n.supportedLocales.');
    }
  }

  if (manifest.i18n?.directory) {
    const directory = manifest.i18n.directory.trim();
    if (!directory) {
      errors.push('i18n.directory cannot be empty.');
    } else if (directory.startsWith('/') || directory.split(/[\\/]/).includes('..')) {
      errors.push('i18n.directory must stay inside the plugin package.');
    }
  }

  if (manifest.compatibility?.appVersion) {
    try {
      satisfiesVersionRange('0.0.0', manifest.compatibility.appVersion);
    } catch (error) {
      errors.push(
        `compatibility.appVersion is invalid: ${error instanceof Error ? error.message : String(error)}`,
      );
    }
  }

  if (manifest.compatibility?.sdkVersion) {
    try {
      satisfiesVersionRange('0.0.0', manifest.compatibility.sdkVersion);
    } catch (error) {
      errors.push(
        `compatibility.sdkVersion is invalid: ${error instanceof Error ? error.message : String(error)}`,
      );
    }
  }

  if (manifest.configFields?.length) {
    const seenKeys = new Set<string>();
    for (const [index, field] of manifest.configFields.entries()) {
      if (field.key?.trim()) {
        if (seenKeys.has(field.key)) {
          errors.push(`configFields contains duplicate key "${field.key}".`);
        }
        seenKeys.add(field.key);
      }
      errors.push(...validateConfigField(field, index));
    }
  }

  return errors;
}

export function validatePluginManifest(manifest: PluginManifest): ManifestValidationResult {
  const errors = getManifestValidationErrors(manifest);
  return {
    valid: errors.length === 0,
    errors,
  };
}

export function createPluginPackageDefinition(
  input: PluginPackageDefinitionInput,
): Record<string, unknown> {
  const main = input.main || 'src/plugin.js';
  const sdkVersion = input.sdkVersion || `^${SDK_VERSION}`;

  return {
    name: input.name,
    version: input.version,
    private: true,
    description: input.description || 'Youwee plugin package',
    type: 'commonjs',
    main,
    scripts: {
      build: 'bunx youwee-sdk build',
      pack: 'bunx youwee-sdk pack --private-key ./plugin.youwee-plugin-key.json',
      keygen: 'bunx youwee-sdk keygen ./plugin.youwee-plugin-key.json',
      'test:deno':
        'deno run --quiet --unstable-detect-cjs --allow-env --allow-read=. --allow-write=. --allow-run node_modules/youwee-sdk/dist/runtime-cli.js src/plugin.js',
    },
    dependencies: {
      'youwee-sdk': sdkVersion,
    },
  };
}

export function createPluginPackageJson(input: PluginPackageDefinitionInput): string {
  return `${JSON.stringify(createPluginPackageDefinition(input), null, 2)}\n`;
}

File: src-tauri/src/services/plugin/manifest.rs
use std::collections::BTreeMap;
use std::path::Path;

use serde_json::Value;

use crate::types::{
    PluginConfigField, PluginConfigFieldInputType, PluginFilesystemPermission, PluginI18nSpec,
    PluginManifest, PluginProvider, PluginRuntimeLanguage,
};

pub(super) fn default_supported_providers(language: &PluginRuntimeLanguage) -> Vec<PluginProvider> {
    match language {
        PluginRuntimeLanguage::Javascript => vec![PluginProvider::Deno],
        PluginRuntimeLanguage::Python => vec![PluginProvider::Python],
    }
}

fn allowed_manifest_triggers() -> &'static [&'static str] {
    &[
        "download.queued",
        "download.beforeStart",
        "download.completed",
        "download.failed",
    ]
}

fn validate_i18n_spec(i18n: &PluginI18nSpec, manifest_path: &Path) -> Result<(), String> {
    if let Some(default_locale) = i18n.default_locale.as_ref() {
        if !i18n.supported_locales.is_empty()
            && !i18n
                .supported_locales
                .iter()
                .any(|locale| locale == default_locale)
        {
            return Err(format!(
                "Plugin manifest {} declares i18n.defaultLocale that is not listed in i18n.supportedLocales",
                manifest_path.display()
            ));
        }
    }

    if let Some(directory) = i18n.directory.as_ref() {
        let path = Path::new(directory);
        if path.is_absolute()
            || path
                .components()
                .any(|component| matches!(component, std::path::Component::ParentDir))
        {
            return Err(format!(
                "Plugin manifest {} declares invalid i18n.directory {}",
                manifest_path.display(),
                directory
            ));
        }
    }

    Ok(())
}

pub(super) fn config_value_matches_field_type(field: &PluginConfigField, value: &Value) -> bool {
    match field.input_type {
        PluginConfigFieldInputType::Text
        | PluginConfigFieldInputType::Textarea
        | PluginConfigFieldInputType::Password
        | PluginConfigFieldInputType::File
        | PluginConfigFieldInputType::Directory
        | PluginConfigFieldInputType::Select => value.is_string(),
        PluginConfigFieldInputType::Number => value.as_f64().is_some(),
        PluginConfigFieldInputType::Boolean => value.is_boolean(),
        PluginConfigFieldInputType::MultiSelect => value
            .as_array()
            .map(|items| items.iter().all(Value::is_string))
            .unwrap_or(false),
    }
}

fn validate_plugin_config_field(
    field: &PluginConfigField,
    manifest_path: &Path,
    field_index: usize,
) -> Result<(), String> {
    if field.key.trim().is_empty() {
        return Err(format!(
            "Plugin manifest {} has a config field at index {} without key",
            manifest_path.display(),
            field_index
        ));
    }
    if field.label.trim().is_empty() {
        return Err(format!(
            "Plugin manifest {} has configFields[{}] without label",
            manifest_path.display(),
            field.key
        ));
    }

    let requires_options = matches!(
        field.input_type,
        PluginConfigFieldInputType::Select | PluginConfigFieldInputType::MultiSelect
    );
    if requires_options && field.options.is_empty() {
        return Err(format!(
            "Plugin manifest {} config field {} must declare options",
            manifest_path.display(),
            field.key
        ));
    }
    if !requires_options && !field.options.is_empty() {
        return Err(format!(
            "Plugin manifest {} config field {} cannot declare options for this inputType",
            manifest_path.display(),
            field.key
        ));
    }

    let mut seen_option_values = BTreeMap::<String, bool>::new();
    for option in &field.options {
        if option.value.trim().is_empty() {
            return Err(format!(
                "Plugin manifest {} config field {} contains an option with empty value",
                manifest_path.display(),
                field.key
            ));
        }
        if option.label.trim().is_empty() {
            return Err(format!(
                "Plugin manifest {} config field {} contains an option with empty label",
                manifest_path.display(),
                field.key
            ));
        }
        if seen_option_values
            .insert(option.value.clone(), true)
            .is_some()
        {
            return Err(format!(
                "Plugin manifest {} config field {} contains duplicate option value {}",
                manifest_path.display(),
                field.key,
                option.value
            ));
        }
    }

    if let Some(default_value) = field.default_value.as_ref() {
        if !config_value_matches_field_type(field, default_value) {
            return Err(format!(
                "Plugin manifest {} config field {} has defaultValue with the wrong type",
                manifest_path.display(),
                field.key
            ));
        }

        if matches!(field.input_type, PluginConfigFieldInputType::Select) {
            let selected = default_value.as_str().unwrap_or_default();
            if !field.options.iter().any(|option| option.value == selected) {
                return Err(format!(
                    "Plugin manifest {} config field {} defaultValue must match a declared option",
                    manifest_path.display(),
                    field.key
                ));
            }
        }

        if matches!(field.input_type, PluginConfigFieldInputType::MultiSelect) {
            let values = default_value.as_array().cloned().unwrap_or_default();
            for value in values {
                let selected = value.as_str().unwrap_or_default();
                if !field.options.iter().any(|option| option.value == selected) {
                    return Err(format!(
                        "Plugin manifest {} config field {} defaultValue contains unsupported option {}",
                        manifest_path.display(),
                        field.key,
                        selected
                    ));
                }
            }
        }
    }

    let uses_number_bounds = field.min.is_some() || field.max.is_some() || field.step.is_some();
    if !matches!(field.input_type, PluginConfigFieldInputType::Number) && uses_number_bounds {
        return Err(format!(
            "Plugin manifest {} config field {} can only use min, max, or step with number inputType",
            manifest_path.display(),
            field.key
        ));
    }
    if let (Some(min), Some(max)) = (field.min, field.max) {
        if min > max {
            return Err(format!(
                "Plugin manifest {} config field {} has min greater than max",
                manifest_path.display(),
                field.key
            ));
        }
    }

    Ok(())
}

pub(super) fn validate_manifest(
    manifest: &PluginManifest,
    manifest_path: &Path,
) -> Result<(), String> {
    if manifest.plugin_id.trim().is_empty() {
        return Err(format!(
            "Plugin manifest {} is missing id",
            manifest_path.display()
        ));
    }
    if manifest.slug.trim().is_empty() {
        return Err(format!(
            "Plugin manifest {} is missing slug",
            manifest_path.display()
        ));
    }
    if manifest.name.trim().is_empty() {
        return Err(format!(
            "Plugin manifest {} is missing name",
            manifest_path.display()
        ));
    }
    if let Some(icon) = manifest.icon.as_ref() {
        if icon.trim().is_empty() {
            return Err(format!(
                "Plugin manifest {} declares an empty icon name",
                manifest_path.display()
            ));
        }
    }
    if manifest.runtime.entrypoint.trim().is_empty() {
        return Err(format!(
            "Plugin manifest {} is missing runtime.entrypoint",
            manifest_path.display()
        ));
    }
    if manifest.runtime.supported_providers.is_empty() {
        return Err(format!(
            "Plugin manifest {} is missing runtime.supportedProviders",
            manifest_path.display()
        ));
    }
    let allowed = default_supported_providers(&manifest.runtime.language);
    for provider in &manifest.runtime.supported_providers {
        if !allowed.iter().any(|candidate| candidate == provider) {
            return Err(format!(
                "Plugin {} declares unsupported provider {} for language {}",
                manifest.plugin_id,
                provider.as_str(),
                manifest.runtime.language.as_str()
            ));
        }
    }
    if let Some(preferred) = manifest.runtime.preferred_provider.as_ref() {
        if !manifest
            .runtime
            .supported_providers
            .iter()
            .any(|provider| provider == preferred)
        {
            return Err(format!(
                "Plugin {} preferredProvider is not listed in supportedProviders",
                manifest.plugin_id
            ));
        }
    }
    let permissions_json = serde_json::to_value(&manifest.permissions)
        .map_err(|e| format!("Failed to inspect plugin permissions: {}", e))?;
    if permissions_json.get("env").is_some() {
        return Err(format!(
            "Plugin manifest {} uses obsolete permissions.env. Define configFields instead.",
            manifest_path.display()
        ));
    }
    if let Some(compatibility) = manifest.compatibility.as_ref() {
        if let Some(range) = compatibility.app_version.as_ref() {
            if range.trim().is_empty() {
                return Err(format!(
                    "Plugin manifest {} has an empty compatibility.appVersion",
                    manifest_path.display()
                ));
            }
        }
        if let Some(range) = compatibility.sdk_version.as_ref() {
            if range.trim().is_empty() {
                return Err(format!(
                    "Plugin manifest {} has an empty compatibility.sdkVersion",
                    manifest_path.display()
                ));
            }
        }
    }
    let mut seen_fs_permissions = BTreeMap::<String, bool>::new();
    for permission in &manifest.permissions.fs {
        if seen_fs_permissions
            .insert(permission.as_str().to_string(), true)
            .is_some()
        {
            return Err(format!(
                "Plugin manifest {} contains duplicate filesystem capability {}",
                manifest_path.display(),
                permission.as_str()
            ));
        }
    }
    let needs_user_selected = manifest.permissions.fs.iter().any(|permission| {
        matches!(
            permission,
            PluginFilesystemPermission::UserSelectedRead
                | PluginFilesystemPermission::UserSelectedWrite
        )
    });
    if needs_user_selected
        && !manifest.config_fields.iter().any(|field| {
            matches!(
                field.input_type,
                PluginConfigFieldInputType::File | PluginConfigFieldInputType::Directory
            )
        })
    {
        return Err(format!(
            "Plugin manifest {} uses fs.user-selected.* but configFields does not declare any file or directory inputs",
            manifest_path.display()
        ));
    }
    if manifest.triggers.is_empty() {
        return Err(format!(
            "Plugin manifest {} must declare at least one trigger",
            manifest_path.display()
        ));
    }
    for trigger in &manifest.triggers {
        if allowed_manifest_triggers()
            .iter()
            .any(|allowed| allowed == trigger)
        {
            continue;
        }
        if trigger.starts_with("triggers.") {
            return Err(format!(
                "Plugin manifest {} contains invalid trigger {}. plugin.json must use raw runtime names like download.completed, not SDK identifiers like triggers.downloadCompleted",
                manifest_path.display(),
                trigger
            ));
        }
        return Err(format!(
            "Plugin manifest {} contains unsupported trigger {}",
            manifest_path.display(),
            trigger
        ));
    }
    if let Some(i18n) = manifest.i18n.as_ref() {
        validate_i18n_spec(i18n, manifest_path)?;
    }
    let mut seen_config_field_keys = BTreeMap::<String, bool>::new();
    for (index, field) in manifest.config_fields.iter().enumerate() {
        if seen_config_field_keys
            .insert(field.key.clone(), true)
            .is_some()
        {
            return Err(format!(
                "Plugin manifest {} contains duplicate config field key {}",
                manifest_path.display(),
                field.key
            ));
        }
        validate_plugin_config_field(field, manifest_path, index)?;
    }
    Ok(())
}

pub(super) fn load_manifest_from_file(manifest_path: &Path) -> Result<PluginManifest, String> {
    let raw = std::fs::read_to_string(manifest_path)
        .map_err(|e| format!("Failed to read {}: {}", manifest_path.display(), e))?;

    let raw_json: Value = serde_json::from_str(&raw)
        .map_err(|e| format!("Failed to parse {}: {}", manifest_path.display(), e))?;
    if raw_json
        .get("permissions")
        .and_then(|value| value.get("env"))
        .is_some()
    {
        return Err(format!(
            "Plugin manifest {} uses obsolete permissions.env. Define configFields instead.",
            manifest_path.display()
        ));
    }

    let manifest: PluginManifest = serde_json::from_str(&raw)
        .map_err(|e| format!("Failed to parse {}: {}", manifest_path.display(), e))?;

    validate_manifest(&manifest, &manifest_path)?;
    Ok(manifest)
}

pub(super) fn load_source_manifest_from_dir(plugin_root: &Path) -> Result<PluginManifest, String> {
    load_manifest_from_file(&plugin_root.join("plugin.json"))
}

pub(super) fn load_installed_manifest_from_dir(
    plugin_root: &Path,
) -> Result<PluginManifest, String> {
    let packaged_manifest = plugin_root.join("manifest.json");
    if packaged_manifest.exists() {
        return load_manifest_from_file(&packaged_manifest);
    }

    load_source_manifest_from_dir(plugin_root)
}
